monai/transforms/atmostonce/functional.py

```diff
@@ -0,0 +1,295 @@
+from monai.transforms.atmostonce.apply import extents_from_shape, shape_from_extents
+from monai.utils import convert_to_tensor, get_equivalent_dtype, ensure_tuple_rep, look_up_option, \
+    GridSampleMode, GridSamplePadMode, fall_back_tuple, ensure_tuple_size, ensure_tuple, InterpolateMode, NumpyPadMode
+
+from monai.config import DtypeLike
+from monai.utils.mapping_stack import MatrixFactory
+
+
+def spacing(
+    img: torch.Tensor,
+    pixdim: Union[Sequence[float], float],
+    src_pixdim: Union[Sequence[float], float],
+    diagonal: Optional[bool] = False,
+    mode: Optional[Union[InterpolateMode, str]] = InterpolateMode.AREA,
+    padding_mode: Optional[Union[NumpyPadMode, GridSamplePadMode, str]] = NumpyPadMode.EDGE,
+    align_corners: Optional[bool] = False,
+    dtype: Optional[Union[DtypeLike, torch.dtype]] = None
+):
+    """
+    Args:
+        img: channel first array, must have shape: (num_channels, H[, W, ..., ]).
+        mode: {``"nearest"``, ``"nearest-exact"``, ``"linear"``,
+            ``"bilinear"``, ``"bicubic"``, ``"trilinear"``, ``"area"``}
+            The interpolation mode. Defaults to ``self.mode``.
+            See also: https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html
+        align_corners: This only has an effect when mode is
+            'linear', 'bilinear', 'bicubic' or 'trilinear'. Defaults to ``self.align_corners``.
+            See also: https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html
+        anti_aliasing: bool, optional
+            Whether to apply a Gaussian filter to smooth the image prior
+            to downsampling. It is crucial to filter when downsampling
+            the image to avoid aliasing artifacts. See also ``skimage.transform.resize``
+        anti_aliasing_sigma: {float, tuple of floats}, optional
+            Standard deviation for Gaussian filtering used when anti-aliasing.
+            By default, this value is chosen as (s - 1) / 2 where s is the
+            downsampling factor, where s > 1. For the up-size case, s < 1, no
+            anti-aliasing is performed prior to rescaling.
+
+    Raises:
+        ValueError: When ``self.spatial_size`` length is less than ``img`` spatial dimensions.
+
+    """
+
+    img_ = convert_to_tensor(img, track_meta=get_track_meta())
+    input_ndim = len(img.shape) - 1
+
+    pixdim_ = ensure_tuple_rep(pixdim, input_ndim)
+    src_pixdim_ = ensure_tuple_rep(src_pixdim, input_ndim)
+
+    if diagonal is True:
+        raise ValueError("'diagonal' value of True is not currently supported")
+
+    mode_ = look_up_option(mode, GridSampleMode)
+    padding_mode_ = look_up_option(padding_mode, GridSamplePadMode)
+    dtype_ = get_equivalent_dtype(dtype or img.dtype, torch.Tensor)
+    zoom_factors = [i / j for i, j in zip(src_pixdim_, pixdim_)]
+
+    transform = MatrixFactory.from_tensor(img).scale(zoom_factors)
+    im_extents = extents_from_shape(img.shape)
```

Comment:
when we have a chain of lazy transforms, `img.shape` will become outdated, correct? because we don't actually run `img_t = apply(img_t)`. 

so I think accessing `img.shape` at this point is inaccurate, this should be something like `metadata['spatial_shape']` where `metadata` should come from a previous lazy transform or `img.shape` if the previous transform evaluates eagerly.

Reply:
and it seems `img` is not really needed in this function (and in the other functionals as well), just need the spatial shape information of it.

Reply:
Yep, I'm aware of the issue (chain of lazy transforms). I'm still trying to figure out the best way to handle when / how transforms to shape / extents occur. I'll hopefully have it sorted by the end of the day; just experimenting at the moment.

Reply:
I think that we should be able describe the extent change by concatenating the individual extent change operations at the point we perform the stacking of transforms; details to follow

Reply:
I've gone the way @wyli went in PR #4911. I thought I could do this by applying the accumulated transform to the extents as they are defined at that point in the chain, but there are a couple of scenarios that seem to break that.